databuilder/dsl.py

```diff
@@ -315,43 +319,80 @@ def _validate_datestring(datestring):
-        return BoolSeries(value=self.value < self._get_other_value(other))
+        return BoolSeries(value=self.value < self._get_other_date_value(other))
 
     def __le__(self, other: DateSeries | str) -> BoolSeries:
-        return BoolSeries(value=self.value <= self._get_other_value(other))
+        return BoolSeries(value=self.value <= self._get_other_date_value(other))
 
     def __ne__(self, other: DateSeries | str) -> BoolSeries:  # type: ignore[override]
-        return BoolSeries(value=self.value != self._get_other_value(other))
+        return BoolSeries(value=self.value != self._get_other_date_value(other))
 
     def round_to_first_of_month(self) -> DateSeries:
         return DateSeries(RoundToFirstOfMonth(self.value))
 
     def round_to_first_of_year(self) -> DateSeries:
         return DateSeries(RoundToFirstOfYear(self.value))
 
-    def __sub__(self, other: str | DateSeries) -> DateDeltaSeries:
-        return DateDeltaSeries(DateDifference(self._get_other_value(other), self.value))
+    def _get_other_datedelta_value(self, delta_value, operation):
+        """Ensure we have either a simple int, or an IntSeries representing a date difference in days"""
+        if isinstance(delta_value, DateDeltaSeries) and isinstance(
+            delta_value.value, DateDifference
+        ):
+            return delta_value.convert_to_days()
+        elif isinstance(delta_value, int):
+            return delta_value
+        else:
+            raise ValueError(
+                f"Can only {operation} integer or DateDeltaSeries (got <{delta_value.__class__.__name__}>)"
+            )
+
+    def __sub__(
+        self, other: str | DateSeries | DateDeltaSeries | IntSeries | int
+    ) -> DateDeltaSeries | DateSeries:
+        if isinstance(other, (str, DateSeries)):
+            return DateDeltaSeries(
+                DateDifference(self._get_other_date_value(other), self.value)
+            )
+        other_value = self._get_other_datedelta_value(other, "subtract")
+        return DateSeries(DateSubtraction(self.value, other_value))
 
     def __rsub__(self, other: str | DateSeries) -> DateDeltaSeries:
-        return DateDeltaSeries(DateDifference(self.value, self._get_other_value(other)))
+        # consistent with python datetime/timedelta, we cannot subtract a DateSeries from
+        # anything other than a date or another DateSeries
+        try:
+            return DateDeltaSeries(
+                DateDifference(self.value, self._get_other_date_value(other))
+            )
+        except ValueError:
+            if isinstance(other, str):
+                raise
+            raise TypeError(
+                f"Can't subtract DateSeries from {other.__class__.__name__}"
+            )
+
+    def __add__(self, other: DateDeltaSeries | int) -> DateSeries:
+        other_value = self._get_other_datedelta_value(other, "add")
+        return DateSeries(DateAddition(self.value, other_value))
```

Comment:
What error would the user see if they try `10 + date_of_birth`?

I'm weakly in favour of supporting it (and ignoring mypy), in order to be mathematically consistent (since we usually expect `A + B` to be `B + A`) but I suspect it's too much work to be worth doing now.

Reply:
It would be an `unsupported operand between...` message

It works fine if I just add an `__radd__` (it has to allow other to be int or DateSeries, and if it's a DateSeries, return other + self so that it uses the DateSeries `__add__` method).  I had that before and decided to take it out because I couldn't find a way to appease mypy and I wasn't sure about just ignoring it.  I could put that back in with the ignore, or we could just leave it till later